pkg/lease/lease.go

```diff
@@ -0,0 +1,82 @@
+package lease
+
+import (
+	"os"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/open-cluster-management/search-collector/pkg/config"
+	coordinationv1 "k8s.io/api/coordination/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+var (
+	agentLabel = map[string]string{"app": "search"}
+)
+
+// LeaseReconciler reconciles a Secret object
+type LeaseReconciler struct {
+	KubeClient           kubernetes.Interface
+	LeaseName            string
+	LeaseDurationSeconds int32
+	componentNamespace   string
+}
+
+func (r *LeaseReconciler) Reconcile() {
+	glog.Info("In lease Reconcile")
+	if len(r.componentNamespace) == 0 {
+		r.componentNamespace = getPodNamespace()
+	}
+	glog.Info("Lease namespace is ", r.componentNamespace)
+	lease, err := config.GetKubeClient().CoordinationV1().Leases(r.componentNamespace).Get(r.LeaseName, metav1.GetOptions{})
+
+	switch {
+	case errors.IsNotFound(err):
+		// create lease
+		lease := &coordinationv1.Lease{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      r.LeaseName,
+				Namespace: r.componentNamespace,
+			},
+			Spec: coordinationv1.LeaseSpec{
+				LeaseDurationSeconds: &r.LeaseDurationSeconds,
+				RenewTime: &metav1.MicroTime{
+					Time: time.Now(),
+				},
+			},
+		}
+		if _, err := r.KubeClient.CoordinationV1().Leases(r.componentNamespace).Create(lease); err != nil {
+			glog.Errorf("unable to create addon lease %q/%q on local managed cluster. error:%v", r.componentNamespace, r.LeaseName, err)
+		} else {
+			glog.Infof("addon lease %q/%q on local managed cluster created", r.componentNamespace, r.LeaseName)
+		}
+
+		return
+	case err != nil:
+		glog.Errorf("unable to get addon lease %q/%q on local managed cluster. error:%v", r.componentNamespace, r.LeaseName, err)
+
+		return
+	default:
+		// update lease
+		lease.Spec.RenewTime = &metav1.MicroTime{Time: time.Now()}
+		if _, err = r.KubeClient.CoordinationV1().Leases(r.componentNamespace).Update(lease); err != nil {
+			glog.Errorf("unable to update cluster lease %q/%q on local managed cluster. error:%v", r.componentNamespace, r.LeaseName, err)
+		} else {
+			glog.Infof("addon lease %q/%q on local managed cluster updated", r.componentNamespace, r.LeaseName)
+		}
+		glog.Info("Exiting lease Reconcile")
+
+		return
+	}
+}
+
+func getPodNamespace() string {
```

Comment:
The value of creating a separate package in our project is exemplified by this method. While it was previously recommended to put the lease creation logic in our operator, our current implementation disqualifies this option. It certainly would be possible, but would create a spaghetti code monster. 

By creating a separate function, we achieve functionality exposed by the operators that does not exist at the search-collector. Example, in the operator, we could just call a function that gives us the namespace, which is required for creating the lease. However, here, we need this method. 

Very clean way of doing this. I'd consider a name other than "lease" to avoid future confusion, but tbh i'm happy with 'lease' for now as it shouldn't cause any code-breaking collisions.